# Исследование объявлений о продаже квартир

**Описание**

В вашем распоряжении данные архива объявлений о продаже квартир в Санкт-Петербурге и соседних населённых пунктов за несколько лет.

По каждой квартире на продажу доступны два вида данных. Первые вписаны пользователем, вторые — получены автоматически на основе картографических данных. Например, расстояние до центра, аэропорта, ближайшего парка и водоёма.

Установление зависимостей позволят построить автоматизированную систему для поиска аномалий и мошеннической деятельности.


**Цель**

Выполнить предобработку данных и предварительный анализ, чтобы зависимости, которые существуют на рынке недвижимости.



**Описание данных**

* `airports_nearest` — расстояние до ближайшего аэропорта в метрах (м)
* `balcony` — число балконов
* `ceiling_height` — высота потолков (м)
* `cityCenters_nearest` — расстояние до центра города (м)
* `days_exposition` — сколько дней было размещено объявление (от публикации до снятия)
* `first_day_exposition` — дата публикации
* `floor` — этаж
* `floors_total` — всего этажей в доме
* `is_apartment` — апартаменты (булев тип)
* `kitchen_area` — площадь кухни в квадратных метрах (м²)
* `last_price` — цена на момент снятия с публикации
* `living_area` — жилая площадь в квадратных метрах (м²)
* `locality_name` — название населённого пункта
* `open_plan` — свободная планировка (булев тип)
* `parks_around3000` — число парков в радиусе 3 км
* `parks_nearest` — расстояние до ближайшего парка (м)
* `ponds_around3000` — число водоёмов в радиусе 3 км
* `ponds_nearest` — расстояние до ближайшего водоёма (м)
* `rooms` — число комнат
* `studio` — квартира-студия (булев тип)
* `total_area` — общая площадь квартиры в квадратных метрах (м²)
* `total_images` — число фотографий квартиры в объявлении

**План работы**

1. Изучить общую информацию о данных
2. Выполнить предобработку входных данных
3. Посчитать и добавить в таблицу новые столбцы:
  * `square_meter_price` - цена одного квадратного метра
  *  `weekday_exposition` - день публикации объявления (`0` - понедельник, `1` - вторник и т.д.)
  * `month_exposition` - месяц публикации объявления
  * `year_exposition` - год публикации объявления
  * `floor_type` - тип этажа квартиры (значения — `первый`, `последний`, `другой`)
  * `distance_center_km` - расстояние в км до центра города
4. Провести исследовательский анализ данных
5. Сформулировать выводы

Литература:
* [Прикладной анализ данных в социальных науках](https://academy.yandex.ru/handbook/data-analysis)
* [Основы Python](https://academy.yandex.ru/handbook/python)
* [Документация Python](https://pythonworld.ru/samouchitel-python)
* [Документация Pandas](https://pandas.pydata.org/docs/user_guide/index.html#user-guide)

## Откройте файл с данными и изучите общую информацию.

In [1]:
import numpy as np
import pandas as pd

In [2]:
data = pd.read_csv(
    'https://code.s3.yandex.net/datasets/real_estate_data.csv',
    sep='\t')

In [3]:
data.head()

,total_images,last_price,total_area,first_day_exposition,rooms,ceiling_height,floors_total,living_area,floor,is_apartment,...,kitchen_area,balcony,locality_name,airports_nearest,cityCenters_nearest,parks_around3000,parks_nearest,ponds_around3000,ponds_nearest,days_exposition
0,20,13000000.0,108.0,2019-03-07T00:00:00,3,2.70,16.0,51.0,8,NaN,...,25.0,NaN,Санкт-Петербург,18863.0,16028.0,1.0,482.0,2.0,755.0,NaN
1,7,3350000.0,40.4,2018-12-04T00:00:00,1,NaN,11.0,18.6,1,NaN,...,11.0,2.0,посёлок Шушары,12817.0,18603.0,0.0,NaN,0.0,NaN,81.0
2,10,5196000.0,56.0,2015-08-20T00:00:00,2,NaN,5.0,34.3,4,NaN,...,8.3,0.0,Санкт-Петербург,21741.0,13933.0,1.0,90.0,2.0,574.0,558.0
3,0,64900000.0,159.0,2015-07-24T00:00:00,3,NaN,14.0,NaN,9,NaN,...,NaN,0.0,Санкт-Петербург,28098.0,6800.0,2.0,84.0,3.0,234.0,424.0
4,2,10000000.0,100.0,2018-06-19T00:00:00,2,3.03,14.0,32.0,13,NaN,...,41.0,NaN,Санкт-Петербург,31856.0,8098.0,2.0,112.0,1.0,48.0,121.0


In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   total_images          23699 non-null  int64  
 1   last_price            23699 non-null  float64
 2   total_area            23699 non-null  float64
 3   first_day_exposition  23699 non-null  object 
 4   rooms                 23699 non-null  int64  
 5   ceiling_height        14504 non-null  float64
 6   floors_total          23613 non-null  float64
 7   living_area           21796 non-null  float64
 8   floor                 23699 non-null  int64  
 9   is_apartment          2775 non-null   object 
 10  studio                23699 non-null  bool   
 11  open_plan             23699 non-null  bool   
 12  kitchen_area          21421 non-null  float64
 13  balcony               12180 non-null  float64
 14  locality_name         23650 non-null  object 
 15  airports_nearest   

In [5]:
data.isna().sum()

,0
total_images,0
last_price,0
total_area,0
first_day_exposition,0
rooms,0
ceiling_height,9195
floors_total,86
living_area,1903
floor,0
is_apartment,20924


In [6]:
data.describe()

,total_images,last_price,total_area,rooms,ceiling_height,floors_total,living_area,floor,kitchen_area,balcony,airports_nearest,cityCenters_nearest,parks_around3000,parks_nearest,ponds_around3000,ponds_nearest,days_exposition
count,23699.000000,2.369900e+04,23699.000000,23699.000000,14504.000000,23613.000000,21796.000000,23699.000000,21421.000000,12180.000000,18157.000000,18180.000000,18181.000000,8079.000000,18181.000000,9110.000000,20518.000000
mean,9.858475,6.541549e+06,60.348651,2.070636,2.771499,10.673824,34.457852,5.892358,10.569807,1.150082,28793.672193,14191.277833,0.611408,490.804555,0.770255,517.980900,180.888634
std,5.682529,1.088701e+07,35.654083,1.078405,1.261056,6.597173,22.030445,4.885249,5.905438,1.071300,12630.880622,8608.386210,0.802074,342.317995,0.938346,277.720643,219.727988
min,0.000000,1.219000e+04,12.000000,0.000000,1.000000,1.000000,2.000000,1.000000,1.300000,0.000000,0.000000,181.000000,0.000000,1.000000,0.000000,13.000000,1.000000
25%,6.000000,3.400000e+06,40.000000,1.000000,2.520000,5.000000,18.600000,2.000000,7.000000,0.000000,18585.000000,9238.000000,0.000000,288.000000,0.000000,294.000000,45.000000
50%,9.000000,4.650000e+06,52.000000,2.000000,2.650000,9.000000,30.000000,4.000000,9.100000,1.000000,26726.000000,13098.500000,0.000000,455.000000,1.000000,502.000000,95.000000
75%,14.000000,6.800000e+06,69.900000,3.000000,2.800000,16.000000,42.300000,8.000000,12.000000,2.000000,37273.000000,16293.000000,1.000000,612.000000,1.000000,729.000000,232.000000
max,50.000000,7.630000e+08,900.000000,19.000000,100.000000,60.000000,409.700000,33.000000,112.000000,5.000000,84869.000000,65968.000000,3.000000,3190.000000,3.000000,1344.000000,1580.000000


In [7]:
data.duplicated().sum()

np.int64(0)

## Предобработка данных

In [8]:
data['first_day_exposition'] = pd.to_datetime(data['first_day_exposition'])
data['is_apartment'] = data['is_apartment'].astype('boolean')
data['locality_name'] = data['locality_name'].astype('string')

In [9]:
list(data['locality_name'].unique())[:10], len(data['locality_name'].unique())

(['Санкт-Петербург',
  'посёлок Шушары',
  'городской посёлок Янино-1',
  'посёлок Парголово',
  'посёлок Мурино',
  'Ломоносов',
  'Сертолово',
  'Петергоф',
  'Пушкин',
  'деревня Кудрово'],
 365)

In [10]:
def clean_loc(name):
    if pd.isna(name):
        return name
    full_name = name.split()
    capitalized = [w for w in full_name if w and (w[0].isupper() or w[0].isdigit())]
    return " ".join(capitalized)

data['locality_name'] = data['locality_name'].apply(clean_loc)
data['locality_name'] = data['locality_name'].astype('string')
len(data['locality_name'].unique())

306

In [11]:
np.sort(data['ceiling_height'].unique())

array([  1.  ,   1.2 ,   1.75,   2.  ,   2.2 ,   2.25,   2.3 ,   2.34,
         2.4 ,   2.45,   2.46,   2.47,   2.48,   2.49,   2.5 ,   2.51,
         2.52,   2.53,   2.54,   2.55,   2.56,   2.57,   2.58,   2.59,
         2.6 ,   2.61,   2.62,   2.63,   2.64,   2.65,   2.66,   2.67,
         2.68,   2.69,   2.7 ,   2.71,   2.72,   2.73,   2.74,   2.75,
         2.76,   2.77,   2.78,   2.79,   2.8 ,   2.81,   2.82,   2.83,
         2.84,   2.85,   2.86,   2.87,   2.88,   2.89,   2.9 ,   2.91,
         2.92,   2.93,   2.94,   2.95,   2.96,   2.97,   2.98,   2.99,
         3.  ,   3.01,   3.02,   3.03,   3.04,   3.05,   3.06,   3.07,
         3.08,   3.09,   3.1 ,   3.11,   3.12,   3.13,   3.14,   3.15,
         3.16,   3.17,   3.18,   3.2 ,   3.21,   3.22,   3.23,   3.24,
         3.25,   3.26,   3.27,   3.28,   3.29,   3.3 ,   3.31,   3.32,
         3.33,   3.34,   3.35,   3.36,   3.37,   3.38,   3.39,   3.4 ,
         3.42,   3.43,   3.44,   3.45,   3.46,   3.47,   3.48,   3.49,
      

In [12]:
def clean_ceiling(h):
    if pd.isna(h):
        return h
    if h < 2 or h == 100:
        return np.nan
    if h >= 20:
        return h / 10
    if h > 6:
        return np.nan
    return h

data['ceiling_height'] = data['ceiling_height'].apply(clean_ceiling)
np.sort(data['ceiling_height'].unique())

array([2.  , 2.2 , 2.25, 2.26, 2.3 , 2.34, 2.4 , 2.45, 2.46, 2.47, 2.48,
       2.49, 2.5 , 2.51, 2.52, 2.53, 2.54, 2.55, 2.56, 2.57, 2.58, 2.59,
       2.6 , 2.61, 2.62, 2.63, 2.64, 2.65, 2.66, 2.67, 2.68, 2.69, 2.7 ,
       2.71, 2.72, 2.73, 2.74, 2.75, 2.76, 2.77, 2.78, 2.79, 2.8 , 2.81,
       2.82, 2.83, 2.84, 2.85, 2.86, 2.87, 2.88, 2.89, 2.9 , 2.91, 2.92,
       2.93, 2.94, 2.95, 2.96, 2.97, 2.98, 2.99, 3.  , 3.01, 3.02, 3.03,
       3.04, 3.05, 3.06, 3.07, 3.08, 3.09, 3.1 , 3.11, 3.12, 3.13, 3.14,
       3.15, 3.16, 3.17, 3.18, 3.2 , 3.21, 3.22, 3.23, 3.24, 3.25, 3.26,
       3.27, 3.28, 3.29, 3.3 , 3.31, 3.32, 3.33, 3.34, 3.35, 3.36, 3.37,
       3.38, 3.39, 3.4 , 3.42, 3.43, 3.44, 3.45, 3.46, 3.47, 3.48, 3.49,
       3.5 , 3.51, 3.52, 3.53, 3.54, 3.55, 3.56, 3.57, 3.58, 3.59, 3.6 ,
       3.62, 3.63, 3.65, 3.66, 3.67, 3.68, 3.69, 3.7 , 3.75, 3.76, 3.78,
       3.8 , 3.82, 3.83, 3.84, 3.85, 3.86, 3.87, 3.88, 3.9 , 3.93, 3.95,
       3.98, 4.  , 4.06, 4.1 , 4.14, 4.15, 4.19, 4.

In [13]:
data['balcony'] = data['balcony'].fillna(0).astype('int64')
data['parks_around3000'] = data['parks_around3000'].fillna(0).astype('int64')
data['ponds_around3000'] = data['ponds_around3000'].fillna(0).astype('int64')

data['is_apartment'] = data['is_apartment'].fillna(False)
data['is_apartment'] = data['is_apartment'].astype('bool')

In [14]:
kitchen_fraction = (data['kitchen_area'] / data['total_area']).median()
living_fraction = (data['living_area'] / data['total_area']).median()

data.loc[data['kitchen_area'].isna(), 'kitchen_area'] = data['total_area'] * kitchen_fraction
data.loc[data['living_area'].isna(), 'living_area'] = data['total_area'] * living_fraction

kitchen_fraction, living_fraction

(0.1724137931034483, 0.5678598992678503)

In [15]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 22 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   total_images          23699 non-null  int64         
 1   last_price            23699 non-null  float64       
 2   total_area            23699 non-null  float64       
 3   first_day_exposition  23699 non-null  datetime64[ns]
 4   rooms                 23699 non-null  int64         
 5   ceiling_height        14494 non-null  float64       
 6   floors_total          23613 non-null  float64       
 7   living_area           23699 non-null  float64       
 8   floor                 23699 non-null  int64         
 9   is_apartment          23699 non-null  bool          
 10  studio                23699 non-null  bool          
 11  open_plan             23699 non-null  bool          
 12  kitchen_area          23699 non-null  float64       
 13  balcony         

## Посчитайте и добавьте в таблицу новые столбцы


In [16]:
data['square_meter_price'] = data['last_price'] / data['total_area']
data['distance_center_km'] = data['cityCenters_nearest'] / 1000

data['weekday_exposition'] = data['first_day_exposition'].dt.weekday
data['month_exposition'] = data['first_day_exposition'].dt.month
data['year_exposition'] = data['first_day_exposition'].dt.year

def floor_type(row):
    f, ft = row['floor'], row['floors_total']
    if f == 1:
        return 'первый'
    if f == ft:
        return 'последний'
    return 'другой'

data['floor_type'] = data[['floor','floors_total']].apply(floor_type, axis=1)

data[['square_meter_price', 'distance_center_km', 'weekday_exposition',
      'month_exposition', 'year_exposition', 'floor_type']].head()

,square_meter_price,distance_center_km,weekday_exposition,month_exposition,year_exposition,floor_type
0,120370.370370,16.028,3,3,2019,другой
1,82920.792079,18.603,1,12,2018,первый
2,92785.714286,13.933,3,8,2015,другой
3,408176.100629,6.800,4,7,2015,другой
4,100000.000000,8.098,1,6,2018,другой


In [17]:
data['floor_type'] = data['floor_type'].astype('string')
data['weekday_exposition'] = data['weekday_exposition'].astype('int64')
data['month_exposition'] = data['month_exposition'].astype('int64')
data['year_exposition'] = data['year_exposition'].astype('int64')
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23699 entries, 0 to 23698
Data columns (total 28 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   total_images          23699 non-null  int64         
 1   last_price            23699 non-null  float64       
 2   total_area            23699 non-null  float64       
 3   first_day_exposition  23699 non-null  datetime64[ns]
 4   rooms                 23699 non-null  int64         
 5   ceiling_height        14494 non-null  float64       
 6   floors_total          23613 non-null  float64       
 7   living_area           23699 non-null  float64       
 8   floor                 23699 non-null  int64         
 9   is_apartment          23699 non-null  bool          
 10  studio                23699 non-null  bool          
 11  open_plan             23699 non-null  bool          
 12  kitchen_area          23699 non-null  float64       
 13  balcony         

## Проведите исследовательский анализ данных

In [18]:
q1, q3 = data['days_exposition'].dropna().quantile([0.25, 0.75])
q1, q3

(45.00000000000001, 232.0)

In [19]:
data[['last_price','total_area','living_area','rooms']].corr()['last_price']

,last_price
last_price,1.000000
total_area,0.653675
living_area,0.585725
rooms,0.363343


In [20]:
top10 = data['locality_name'].value_counts().head(10).index
avg_price = (data[data['locality_name'].isin(top10)]
             .groupby('locality_name')['square_meter_price']
             .mean().sort_values(ascending=False))
avg_price

,square_meter_price
locality_name,
Санкт-Петербург,114849.008794
Пушкин,103125.819291
Кудрово,95324.930544
Парголово,90175.913458
Мурино,86087.512050
Шушары,78677.364217
Колпино,75424.579098
Гатчина,68746.146763
Всеволожск,68654.473986


In [21]:
data[data['locality_name'] == 'Санкт-Петербург'][['last_price', 'distance_center_km']].corr()['last_price']

,last_price
last_price,1.000000
distance_center_km,-0.259507


## Общий вывод

В ходе работы была изучена общая информация о данных, выполнена их предобработка, произведены и добавлены в таблицу новые столбцы.
Исследовательский анализ данных позволил установить время продажи квартир, выявить факторы, оказывающие влияние на общую (полную) стоимость объекта.

Примерное время продажи квартир находится в диапазоне от `45` до `232` дней. Общая (полная) стоимость объекта коррелирует и имеет положитеьную связь с общей площадью, жилой площадью и количеством комнат. Было установлено, что средняя цена одного квадратного метра в 10 населённых пунктах с наибольшим числом объявлений показывает, что самая высокая стоимость квадратного метра в `Санкт-Петербург`, а самая низка - в `Выборг`. Также было установлено, что с удалением от центра `Санкт-Петербург` средняя цена за объект в этом городе `уменьшается` с каждым киллометром, что подтверждается `отрицательной корреляцией -0,26` между рассматриваемыми переменными.